finder_server.cc

```diff
@@ -0,0 +1,237 @@
+}
+
+
+InventoryInfo VendorClient::InquireInventoryInfo (uint32_t food_id) {
+  FoodID request;
+  request.set_food_id(food_id);
+  
+  InventoryInfo info;
+  
+  ClientContext context;
+  
+  Status status = vendor_stub_->CheckInventory(&context, request, &info);
+  
+  if (status.ok()) {
+    return info;
+  } else {
+    std::cout << status.error_code() << ": " << status.error_message()
+              << std::endl;
+    info.set_price(-1);
+    return info;
+  }
+}
+
+
+string SupplierClient::SayHelloToSupplier(const string& user) {
+  // Data we are sending to the server.
+  HelloRequest request;
+  request.set_name(user);
+
+  HelloReply reply;
+
+  ClientContext context;
+
+  // The actual RPC.
+  Status status = supplier_stub_->SayHello(&context, request, &reply);
+
+  // Act upon its status.
+  if (status.ok()) {
+    return reply.message();
+  } else {
+    std::cout << status.error_code() << ": " << status.error_message()
+              << std::endl;
+    return "RPC failed";
+  }
+}
+
+vector<VendorInfo> SupplierClient::InquireVendorInfo (uint32_t food_id) {
+  FoodID request;
+  request.set_food_id(food_id);
+  
+  VendorInfo reply;
+  vector<VendorInfo> info;
+  
+  ClientContext context;
+  std::unique_ptr<ClientReader<VendorInfo>> reader (supplier_stub_->CheckVendor(&context, request));
+  // Status status = supplier_stub_->CheckVendor(&context, request, &info);
+
+  while (reader->Read(&reply)) {
+    info.push_back(reply);
+  }
+
+  Status status = reader->Finish();
+  
+  if (status.ok()) {
+    return info;
+  } else {
+    std::cout << status.error_code() << ": " << status.error_message()
+              << std::endl;
+    return info;
+  }
+}
+
+void FinderServiceImpl::PrintVendorInfo (const uint32_t id, const vector<VendorInfo>& info) {
+  std::cout << "The following vendors might have food ID " << id << std::endl;
+  for (auto & i : info) {
+    std::cout << "\tVendor url: " << i.url() << "; name: " << i.name()
+      << "; location: " << i.location() << std::endl;
+  }
+}
+
+vector<ShopInfo> FinderServiceImpl::ProcessRequest(uint32_t food_id) {
+  /*
+   * Initiate client with supplier channel. Retrieve a list of vendor address
+   * Then create client
+   * return a vector of <Vendor Info, Inventory Info>
+   */
+  vector<ShopInfo> result;
+        
+  vector<VendorInfo> vendor_info = supplier_client_.InquireVendorInfo(food_id);
+  
+  FinderServiceImpl::PrintVendorInfo(food_id, vendor_info);
+  
+  for (const auto & vendor : vendor_info) {
+    // TODO: save these clients 
+    VendorClient vendor_client(grpc::CreateChannel(
+      vendor.url(), grpc::InsecureChannelCredentials()));
+    InventoryInfo inventory_info = vendor_client.InquireInventoryInfo(food_id);
+    // error checking: if no inventory, price == -1
+    if (inventory_info.price() < 0)
+      std::cout << "vendor at " << vendor.url() << " doesn't have food " << food_id << std::endl;
+    else {
+      VendorInfo* v_ptr = new VendorInfo(vendor);
```

Comment:
Instead of using `new`, you should use `info.mutable_vendor()` to get an already allocated pointer that you can fill in (see https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#embeddedmessage).  Avoid using `new` in general.